rln/src/merkle_tree.rs

```diff
@@ -1,251 +1,231 @@
+// Implementation adapted from https://github.com/kilic/rln/blob/master/src/merkle.rs
+// and https://github.com/worldcoin/semaphore-rs/blob/d462a4372f1fd9c27610f2acfe4841fab1d396aa/src/merkle_tree.rs
 
-//! Implements basic binary Merkle trees
 //!
 //! # To do
 //!
 //! * Disk based storage backend (using mmaped files should be easy)
 
-use semaphore::Field;
+use ark_std::str::FromStr;
 use serde::{Deserialize, Serialize};
-use std::{
-    cmp::max,
-    fmt::Debug,
-    iter::{once, repeat, successors},
-};
+use std::cmp::max;
+use std::fmt::Debug;
+use std::io::{self, Error, ErrorKind};
+use std::{collections::HashMap, hash::Hash};
 
-/// Hash types, values and algorithms for a Merkle tree
+/// In Hasher we define the node type, the default leaf and the hash function used to initialize the Merkle Tree
 pub trait Hasher {
-    /// Type of the leaf and node hashes
-    type Hash: Copy + Clone + Eq + Serialize;
+    /// Type of the leaf and tree node
+    type Fr: Copy + Clone + Eq + Serialize;
 
-    /// Compute the hash of an intermediate node
-    fn hash_node(left: &Self::Hash, right: &Self::Hash) -> Self::Hash;
-}
-
-/// Merkle tree with all leaf and intermediate hashes stored
-#[derive(Clone, PartialEq, Eq, Debug)]
-pub struct MerkleTree<H: Hasher> {
-    /// total number of levels of the tree, i.e. # of layers including pre-images layer of tree leaves
-    levels: usize,
-
-    /// Hash value of empty subtrees of given levels, starting at leaf level
-    empty: Vec<H::Hash>,
+    /// Returns the default tree leaf
+    fn default_leaf() -> Self::Fr;
 
-    /// Hash values of tree nodes and leaves, breadth first order
-    nodes: Vec<H::Hash>,
-
-    // The next available (i.e., never used) tree index. Equivalently, the number of leaves added to the tree
-    // (deletions leave next_index unchanged)
-    pub next_index: usize,
+    /// Utility to compute the hash of an intermediate node
+    fn hash(input: &[Self::Fr]) -> Self::Fr;
 }
 
-/// Element of a Merkle proof
-#[derive(Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
-pub enum Branch<H: Hasher> {
-    /// Left branch taken, value is the right sibling hash.
-    Left(H::Hash),
-
-    /// Right branch taken, value is the left sibling hash.
-    Right(H::Hash),
+/// The Merkle tree structure
+pub struct MerkleTree<H>
+where
+    H: Hasher,
+{
+    /// The depth of the tree
+    pub depth: usize,
+    /// The nodes cached from the empty part of the tree (where leaves are set to default).
+    /// Since the rightmost part of the tree is usually changed much later than its creation,
+    /// we can prove accumulation of elements in the leftmost part, with no need to initialize the full tree
+    /// and by caching few intermediate nodes to the root computed from default leaves
+    cached_nodes: Vec<H::Fr>,
```

Comment:
Yes, I know the purpose of the `zero` vector, it essentially a way to represent a `zero` or `empty` Merkle tree with logarithmic storage complexity i.e., by only `d+1` elements. The name `zeros` also  mirrors how it is being called in the literature. It is fine to keep it `cached_nodes`, but sticking to the name `zeros` allows other developers to quickly understand it maps to the actual `zeros` algorithm and nothing different.

Reply:
No strong opinions on above, a meta comment though: try to avoid back and forth in PRs, especially if isn't just asking about clarification but more of a discussion. If a choice is controversial it is better to make the merge the part that people agree on/are OK with, and isolate discussions to specific follow up issues.

Usually, it doesn't matter that much.